src/AggregateFunctions/AggregateFunctionGroupArrayIntersect.cpp

```diff
@@ -0,0 +1,449 @@
+template <typename T>
+class AggregateFunctionGroupArrayIntersect
+    : public IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>, AggregateFunctionGroupArrayIntersect<T>>
+{
+
+private:
+    using State = AggregateFunctionGroupArrayIntersectData<T>;
+
+public:
+    AggregateFunctionGroupArrayIntersect(const DataTypePtr & argument_type, const Array & parameters_)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>,
+          AggregateFunctionGroupArrayIntersect<T>>({argument_type}, parameters_, std::make_shared<DataTypeArray>(argument_type)) {}
+
+    AggregateFunctionGroupArrayIntersect(const DataTypePtr & argument_type, const Array & parameters_, const DataTypePtr & result_type_)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>,
+          AggregateFunctionGroupArrayIntersect<T>>({argument_type}, parameters_, result_type_) {}
+
+    String getName() const override { return "GroupArrayIntersect"; }
+
+    bool allocatesMemoryInArena() const override { return false; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        auto & version = this->data(place).version;
+        auto & map = this->data(place).value;
+        const auto data_column = assert_cast<const ColumnArray &>(*columns[0]).getDataPtr();
+        const auto & offsets = assert_cast<const ColumnArray &>(*columns[0]).getOffsets();
+        const size_t offset = offsets[static_cast<ssize_t>(row_num) - 1];
+        const auto arr_size = offsets[row_num] - offset;
+        ++version;
+        if (version == 1)
+        {
+            for (size_t i = 0; i < arr_size; ++i)
+                map[static_cast<T>((*data_column)[offset + i].get<T>())] = version;
+        }
+        else if (map.size() > 0)
+        {
+            for (size_t i = 0; i < arr_size; ++i)
+            {
+                typename State::Map::LookupResult value = map.find(static_cast<T>((*data_column)[offset + i].get<T>()));
+                if (value != nullptr && value->getMapped() == version - 1)
+                    ++(value->getMapped());
+            }
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena *) const override
+    {
+        auto & map = this->data(place).value;
+        const auto & rhs_map = this->data(rhs).value;
+        UInt64 rhs_version = this->data(rhs).version;
+        if (rhs_version == 0)
+            return;
+        UInt64 version = this->data(place).version++;
+        if (version == 0)
+        {
+            for (auto & rhs_elem : rhs_map)
+            {
+                if (rhs_elem.getMapped() == rhs_version)
+                    map[rhs_elem.getKey()] = 1;
+            }
+            return;
+        }
+        for (auto & rhs_elem : rhs_map)
+        {
+            if (rhs_elem.getMapped() != rhs_version)
+                continue;
```

Comment:
Is data with an old version evicted from memory at any time point?
Consider an example:
``` sql
INSERT INTO test_numbers SELECT range(1000000);
INSERT INTO test_numbers VALUES ([100]);
 ```
 
Only one element has the most recent version, the others are no longer relevant.